### PR TITLE
Fixed #27673 -- Prevented system checks for contrib.admin during registration.

### PR DESCRIPTION
**Pull request by ellmetha**
https://code.djangoproject.com/ticket/27673

### CODE REVIEW

File: django/contrib/admin/checks.py [hunk 1]
```diff
@@ -19,9 +19,11 @@
 
 
 def check_admin_app(**kwargs):
-    from django.contrib.admin.sites import system_check_errors
-
-    return system_check_errors
+    from django.contrib.admin.sites import system_checks
```
Comment (codingjoe) [inline django/contrib/admin/checks.py]: I think it would make more sense to drop the entire method and register all checks separately in the admin's AppConfig

**Comment by codingjoe** [comment]
@ellmetha thanks! Thats great, I really need this :)

**Comment by codingjoe** [comment]
Oh, btw, Django commit messages are written in past tense.

File: django/contrib/admin/checks.py (hunk 1 above)
Comment (ellmetha) [inline django/contrib/admin/checks.py]: I'm not sure about that. Currently the check function (`check_admin_app`) is explicitly defined in the `checks` module ; this seems to be a good solution IMO. :wink: Why would it make more sense to change this?

**Comment by timgraham** [comment]
This similar to what @adamchainz did in d57a4ba1f228f51fe4572a9b01c981bab095dd03 (a commit in https://github.com/django/django/pull/6980). Could you  compare the two approaches?

**Comment by adamchainz** [comment]
N.B. I do intend to go back to that one, its dependency #6294 was only merged last week

**Comment by ellmetha** [comment]
@timgraham well the main difference is that a `Site.check` method is introduced in d57a4ba (which is fine I think but perhaps not necessary). The two approaches are equivalent though.

**Comment by timgraham** [comment]
Adding an `AdminSite.check()` method looks like a good approach to me. That's the pattern we've been moving toward (e.g. https://github.com/django/django/pull/7781 and e7fa89fb58cc7ba468e0167f506dc4fce7ec532a).

**Comment by ellmetha** [comment]
@timgraham I've made the proposed change.

File: django/contrib/admin/sites.py [hunk 2]
```diff
@@ -63,6 +64,21 @@ def __init__(self, name='admin'):
...
+
+    def check(self):
+        """
+        Run the system checks of the underlying admin objects.
+        """
+        if not settings.DEBUG:
+            return []
+
+        errors = []
+        admin_objects_to_check = [o for o in self._registry.values() if o.__class__ is not ModelAdmin]
```
Comment (adamchainz) [inline django/contrib/admin/sites.py]: why did you add this `is not` ?

Comment (ellmetha) [inline django/contrib/admin/sites.py]: Because checks are currently not executed for models whose admin class is `ModelAdmin` (see: https://github.com/django/django/blob/master/django/contrib/admin/sites.py#L109). This is to keep the same behavior. :wink:

Comment (adamchainz) [inline django/contrib/admin/sites.py]: Oh yes. I don't know why that is, it's an odd restriction, I remember now I didn't keep it in #6980 because I didn't understand why it existed.

**Comment by adamchainz** [comment]
I just updated #6980. I think the `AdminSite.check()` implementation there, which is only a little different, is preferable as it also fixes the weird `DEBUG=True` behaviour.

**Comment by adamchainz** [comment]
N.B. We could of course merge this first and rebase #6980 on top of this so you get added as a contributor in the git log @ellmetha 😉

File: django/contrib/admin/sites.py (hunk 2 above)
Comment (codingjoe) [inline django/contrib/admin/sites.py]: I think you should use `type` not `__class__`. At least in python 2 this will break if the class does not inherit from `object`.

Comment (codingjoe) [inline django/contrib/admin/sites.py]: Also this doesn't need to be a list. A generator will do, since you loop only once over the result.

File: tests/admin_checks/tests.py [hunk 3]
```diff
@@ -105,7 +111,30 @@ class CustomAdminSite(admin.AdminSite):
...
+                return errors
+
+        class MyAuthorAdmin(admin.ModelAdmin):
+            pass
+
+        admin.site.register(Book, MyBookAdmin)
+        admin.site.register(Author, MyAuthorAdmin)
+
+        try:
+            errors = checks.run_checks()
```
Comment (codingjoe) [inline tests/admin_checks/tests.py]: The lines below do not belong into the try block, since they don't raise the exception.
```python
try:
    errors = checks.run_checks()
else:
    expected = []
    self.assertEqual(errors, expected)
finally:
    admin.site.unregister(Book)
    admin.site.unregister(Author)
```

File: django/contrib/admin/sites.py (hunk 2 above)
Comment (timgraham) [inline django/contrib/admin/sites.py]: Does that mean something like `admin.site.register(Choice, list_display=['question'])` isn't checked? If so, I that seems like an incorrect restriction.

Comment (ellmetha) [inline django/contrib/admin/sites.py]: @timgraham I don't think so. See: https://github.com/django/django/blob/master/django/contrib/admin/sites.py#L105 If there are options associated. A new admin subclass of `admin_class`  is dynamically created if there are options defined in the `register` call, thus ensuring the checks are executed for this subclass.

Comment (timgraham) [inline django/contrib/admin/sites.py]: So maybe it's a performance optimization? A comment would be useful.

Comment (ellmetha) [inline django/contrib/admin/sites.py]: This line seems to come from 4ad1eb1. I'd say that this does not make sense to validate / execute system checks for a `ModelAdmin` instance if no custom options have been defined during the registration.

**Comment by ellmetha** [comment]
Thanks for your suggestions @codingjoe I've made the proposed changes.

File: tests/admin_checks/tests.py [hunk 4]
```diff
@@ -43,6 +45,11 @@ def check(self, **kwargs):
 )
 class SystemChecksTestCase(SimpleTestCase):
 
+    def setUp(self):
+        super(SystemChecksTestCase, self).setUp()
+        admin.sites.all_sites.clear()
```
Comment (adamchainz) [inline tests/admin_checks/tests.py]: This isn't great and invites test state leakage - it's directly touching the internals of the code under test and not resetting it to the way it was after the tests. It would be best if a copy of `all_sites` was made and then in `tearDown` it was reset to the original state, at least

Possibly fixed by #6980 anyway

Comment (ellmetha) [inline tests/admin_checks/tests.py]: You're right. I fixed this.

File: django/contrib/admin/sites.py (hunk 2 above)
Comment (adamchainz) [inline django/contrib/admin/sites.py]: @ellmetha some checks look at the suitability of the model class for the admin, rather than just attributes on `ModelAdmin`, it's rather presumptive to skip them...

File: tests/admin_checks/tests.py [hunk 5]
```diff
@@ -43,6 +45,17 @@ def check(self, **kwargs):
 )
 class SystemChecksTestCase(SimpleTestCase):
 
+    def setUp(self):
+        super(SystemChecksTestCase, self).setUp()
```
Comment (timgraham) [inline tests/admin_checks/tests.py]: No need to call super in setUp/tearDown

File: tests/admin_checks/tests.py [hunk 6]
```diff
@@ -105,7 +117,29 @@ class CustomAdminSite(admin.AdminSite):
...
+            pass
+
+        admin.site.register(Book, MyBookAdmin)
+        admin.site.register(Author, MyAuthorAdmin)
+
+        errors = checks.run_checks()
+        expected = []
+        self.assertEqual(errors, expected)
+
+        admin.site.unregister(Book)
```
Comment (timgraham) [inline tests/admin_checks/tests.py]: Why did you remove the try/finally?

File: django/contrib/admin/sites.py [hunk 7]
```diff
@@ -63,6 +64,21 @@ def __init__(self, name='admin'):
...
+
+    def check(self):
+        """
+        Run the system checks of the underlying admin objects.
+        """
+        if not settings.DEBUG:
+            return []
+
+        errors = []
+        admin_objects_to_check = (o for o in self._registry.values() if type(o) is not ModelAdmin)
```
Comment (timgraham) [inline django/contrib/admin/sites.py]: I'd keep the current version without `type()` since it's unrelated. If there's a need for this change, create a ticket with a corresponding test.

File: django/contrib/admin/sites.py [hunk 8]
```diff
@@ -63,6 +64,21 @@ def __init__(self, name='admin'):
         self.name = name
         self._actions = {'delete_selected': actions.delete_selected}
         self._global_actions = self._actions.copy()
+        all_sites.add(self)
+
+    def check(self):
+        """
+        Run the system checks of the underlying admin objects.
```
Comment (timgraham) [inline django/contrib/admin/sites.py]: how about: "Run the system checks on all ModelAdmins, except if they aren't customized at all."

File: django/contrib/admin/sites.py (hunk 7 above)
Comment (timgraham) [inline django/contrib/admin/sites.py]: I think that's a good name if the code were more complicated but `modeladmins` and `for modeladmin in modeladmins:` seems okay to me.

File: django/contrib/admin/sites.py [hunk 9]
```diff
@@ -63,6 +64,21 @@ def __init__(self, name='admin'):
...
+        """
+        if not settings.DEBUG:
+            return []
+
+        errors = []
+        admin_objects_to_check = (o for o in self._registry.values() if type(o) is not ModelAdmin)
+        for admin_obj in admin_objects_to_check:
+            errors.extend(admin_obj.check())
+
```
Comment (timgraham) [inline django/contrib/admin/sites.py]: chop blank line

File: django/contrib/admin/sites.py [hunk 10]
```diff
@@ -106,8 +122,6 @@ def register(self, model_or_iterable, admin_class=None, **options):
 
                 # Instantiate the admin class to save in the registry
                 admin_obj = admin_class(model, self)
-                if admin_class is not ModelAdmin and settings.DEBUG:
-                    system_check_errors.extend(admin_obj.check())
 
                 self._registry[model] = admin_obj
```
Comment (timgraham) [inline django/contrib/admin/sites.py]: `self._registry[model] = admin_class(model, self)` looks fine now

File: tests/admin_checks/tests.py [hunk 11]
```diff
@@ -43,6 +45,17 @@ def check(self, **kwargs):
 )
 class SystemChecksTestCase(SimpleTestCase):
 
+    def setUp(self):
+        super(SystemChecksTestCase, self).setUp()
+        self.previous_sites = list(admin.sites.all_sites)
```
Comment (timgraham) [inline tests/admin_checks/tests.py]: We typically use an underscore prefix in a case like this: `self._previous_sites`

File: tests/admin_checks/tests.py [hunk 12]
```diff
@@ -105,7 +117,29 @@ class CustomAdminSite(admin.AdminSite):
...
+                    errors.append('AuthorAdmin missing!')
+                return errors
+
+        class MyAuthorAdmin(admin.ModelAdmin):
+            pass
+
+        admin.site.register(Book, MyBookAdmin)
+        admin.site.register(Author, MyAuthorAdmin)
+
+        errors = checks.run_checks()
```
Comment (timgraham) [inline tests/admin_checks/tests.py]: Would calling `admin.site.check()` work, e.g. `self.assertEqual(admin.site.check(), [])`? That avoids running checks for all installed apps which should be a speed improvement. Other tests could be refactored later.

Comment (ellmetha) [inline tests/admin_checks/tests.py]: This could work. I could make this modification for this test but I think that by calling `checks.run_checks()` you also test that the admin checks are properly registered into the checks registry.

File: django/contrib/admin/sites.py (hunk 8 above)
Comment (ellmetha) [inline django/contrib/admin/sites.py]: Sounds good.

File: tests/admin_checks/tests.py (hunk 12 above)
Comment (timgraham) [inline tests/admin_checks/tests.py]: Yes, we'd need a separate test for check registration. Possibly https://github.com/django/django/pull/7781 will come up with a template to use.

**Comment by ellmetha** [comment]
@timgraham I've made the proposed change. :wink:

File: tests/admin_checks/tests.py (hunk 4 above)
Comment (timgraham) [inline tests/admin_checks/tests.py]: Adam, I'm unsure what you mean -- are you saying you don't think the test is adding value or that the setUp/tearDown is obsoleted by some change in your PR? I understand this PR is basically the same as what you did in 22f027e1bdce3d74ae828710eeaeca78b8c0c170 (with a test added).

Comment (adamchainz) [inline tests/admin_checks/tests.py]: I think the setup/teardown is obsoleted by my PR only running checks for models in installed apps

**Comment by timgraham** [comment]
I incorporated Adam's work from https://github.com/django/django/issues/6980. I think this is good to go.

**Comment by adamchainz** [comment]
🎉 🎉  Thanks @ellmetha, @codingjoe, and @timgraham
